pkg/actuators/machineset/controller.go

```diff
@@ -0,0 +1,144 @@
+	logger.V(3).Info("Reconciling")
+
+	machineSet := &machinev1beta1.MachineSet{}
+	if err := r.Client.Get(ctx, req.NamespacedName, machineSet); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Ignore deleted MachineSets, this can happen when foregroundDeletion
+	// is enabled
+	if !machineSet.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+	originalMachineSetToPatch := client.MergeFrom(machineSet.DeepCopy())
+
+	result, err := reconcile(machineSet)
+	if err != nil {
+		logger.Error(err, "Failed to reconcile MachineSet")
+		r.recorder.Eventf(machineSet, corev1.EventTypeWarning, "ReconcileError", "%v", err)
+		// we don't return here, so we want to attempt to patch the machine regardless of an error.
+	}
+
+	if err := r.Client.Patch(ctx, machineSet, originalMachineSetToPatch); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to patch machineSet: %w", err)
+	}
+
+	if isInvalidConfigurationError(err) {
+		// For situations where requeuing won't help we don't return error.
+		// https://github.com/kubernetes-sigs/controller-runtime/issues/617
+		return result, nil
+	}
+	return result, err
+}
+
+func isInvalidConfigurationError(err error) bool {
+	switch t := err.(type) {
+	case *mapierrors.MachineError:
+		if t.Reason == machinev1beta1.InvalidConfigurationMachineError {
+			return true
+		}
+	}
+	return false
+}
+
+func reconcile(machineSet *machinev1beta1.MachineSet) (ctrl.Result, error) {
+	providerConfig, err := machineactuator.ProviderSpecFromRawExtension(machineSet.Spec.Template.Spec.ProviderSpec.Value)
+	if err != nil {
+		return ctrl.Result{}, mapierrors.InvalidMachineConfiguration("failed to get providerConfig: %v", err)
+	}
+
+	if machineSet.Annotations == nil {
+		machineSet.Annotations = make(map[string]string)
+	}
+	cpu, err := getPowerVSProcessorValue(providerConfig.Processors)
+	if err != nil {
+		return ctrl.Result{}, mapierrors.InvalidMachineConfiguration("failed to get cpu value: %v", err)
+	}
+	machineSet.Annotations[cpuKey] = cpu
+	// Convert Memory from GiB to Mb
+	machineSet.Annotations[memoryKey] = strconv.FormatInt(int64(providerConfig.MemoryGiB)*1024, 10)
+	return ctrl.Result{}, nil
+}
+
+func getPowerVSProcessorValue(processor intstr.IntOrString) (string, error) {
+	var cores float64
+	var virtualProcessors string
+	var err error
+	switch processor.Type {
+	case intstr.Int:
+		cores = float64(processor.IntVal)
+	case intstr.String:
+		if cores, err = strconv.ParseFloat(processor.StrVal, 64); err != nil {
+			return virtualProcessors, err
+		}
+	}
+	// 1 Power VS machine core will be equals to 8 virtual cores
+	virtualProcessors = fmt.Sprintf("%v", math.Ceil(cores)*8)
```

Comment:
Add more information like
```
There is a core-to-vCPU ratio of 1:1. For shared processors, fractional cores round up to the nearest whole number. For example, 1.25 cores equals 2 vCPUs.
```
and the sample output of `lparstat` and `lscpu` command for the reference.

Reply:
is core-to-vCPU ratio is 1:1 or 1:8? because I see CPUS as 8

```
[core@rdr-kn24-f9jtx-master-0 ~]$ lscpu
Architecture:        ppc64le
Byte Order:          Little Endian
CPU(s):              8
On-line CPU(s) list: 0-7
Thread(s) per core:  8
Core(s) per socket:  0
Socket(s):           2
NUMA node(s):        1
Model:               2.2 (pvr 004e 0202)
Model name:          POWER9 (architected), altivec supported
Hypervisor vendor:   pHyp
Virtualization type: para
L1d cache:           32K
L1i cache:           32K
NUMA node1 CPU(s):   0-7
```

Reply:
As we discussed offline, the logic here is:

This is the formula used to get that count in the `CPU(s): ` section in the OS

`Entitlement(rounded) * SMT`

Default SMT is 8

e.g:

VM with 1 dedicated processor will see = 1 * SMT = 1 * 8 = 8 cpus in OS
VM with 1.5 shared processor will see = 2 * SMT = 2 * 8 = 16 cpus in OS

Reply:
Updated it